src/swap/handle_check_address.c

```diff
@@ -0,0 +1,137 @@
+#ifdef HAVE_SWAP
+
+#include <string.h>
+
+#include "os.h"
+#include "crypto_helpers.h"
+#include "swap.h"
+#include "../types.h"
+#include "common/mybuffer.h"
+#include "get_public_key.h"
+#include "base64.h"
+#include "address.h"
+#include "constants.h"
+#include "crc16.h"
+
+#define BASE_CHAIN   0x00
+#define MASTER_CHAIN 0xFF
+
+/* Set params.result to 0 on error, 1 otherwise */
+void swap_handle_check_address(check_address_parameters_t *params) {
+    PRINTF("Inside Ton swap_handle_check_address\n");
+    params->result = 0;
+
+    if (params->address_parameters == NULL) {
+        PRINTF("derivation path expected\n");
+        return;
+    }
+    PRINTF("address_parameters %.*H\n",
+           params->address_parameters_length,
+           params->address_parameters);
+
+    if (params->address_to_check == NULL) {
+        PRINTF("Address to check expected\n");
+        return;
+    }
+    PRINTF("Address to check %s\n", params->address_to_check);
+    if (strlen(params->address_to_check) != 48) {
+        PRINTF("Address to check expected length 48, not %d\n", strlen(params->address_to_check));
+        return;
+    }
+
+    buffer_t cdata;
+    cdata.ptr = params->address_parameters;
+    cdata.size = params->address_parameters_length;
+    cdata.offset = 0;
+
+    uint8_t bip32_path_len;
+    uint32_t bip32_path[MAX_BIP32_PATH];
+    pubkey_ctx_t pk_info;
+    pk_info.subwallet_id = 698983191;
+    pk_info.is_v3r2 = false;
+    if (get_public_key_helper(0, &cdata, &bip32_path_len, bip32_path, &pk_info) != 0) {
+        PRINTF("Failed to read public key\n");
+        return;
+    }
+    PRINTF("Derived on path %.*H\n", 4 * bip32_path_len, bip32_path);
+    PRINTF("Public key %.*H\n", PUBKEY_LEN, pk_info.raw_public_key);
+
+    uint8_t hash[HASH_LEN] = {0};
+    if (!pubkey_to_hash(pk_info.raw_public_key,
+                        pk_info.subwallet_id,
+                        pk_info.is_v3r2,
+                        hash,
+                        sizeof(hash))) {
+        return;
+    }
+    PRINTF("hash %.*H\n", HASH_LEN, hash);
+
+    char address_base64url[48];
```

Comment:
Hardcoded sizes are difficult to understand. A comment would be welcome :wink:

Reply:
👍 define added